### PR TITLE
fix: handle timeout and unexpected eof error by returning 408 or 499

### PR DESCRIPTION
**Pull request by mar4uk**
In case of errors

- unexpected EOF the connection is broken and nothing else can be read (the user has a bad connection, the user closed the request): 499
- read timeout means the connection is there and nothing is being sent - so that would be the disappearing act: user opened the page and gone: 408

we shouldn't return 500, we should return 499 or 408 instead

### CODE REVIEW

**Review by jesusvazquez — COMMENTED** [review]
:wave: Thanks for the contribution Irina, 

I'm happy with most of the PR specially the 408 handling but I don't agree on the 499 usage. 499 is not an official http code so I'd rather not use it.

**Comment by mar4uk** [comment]
Thank you for reviewing @jesusvazquez, yeah, 499 status is not an official status code and is specific to the nginx server.
Do you think it should be 408 status code for unexpected eof error?

**Comment by jesusvazquez** [comment]
I'm not sure what to do with the EOF error. It can come from multiple sources, it could be our own handling of the requests, that we're not doing it right and that leads to EOFs, it can also be the client closing the requests unexpectedly or just a network issue. This is what is giving me trouble because if we reply with a 4XX thats a always a client error.

Are you getting many of these?

**Comment by mar4uk** [comment]
I will tell you a little bit later when we deploy this fix for faro (deployment is currently in progress) and it stays for a while. Then I can more accurately say how many 499 we get. But right now we have 
avg QPS:
1.6K 200
0.3 500 (all 500 are actually 408 or 499)

**Review by jesusvazquez — APPROVED** [review]

**Comment by jesusvazquez** [comment]
Oki, approving for now we can revisit if we're doing the proper handling of that http code later. This way I unblock you :+1:

**Comment by mar4uk** [comment]
We are not blocked by this PR, we did a fork of mimir-proxies with this fix.
We can keep this PR open and add some improvements if they are needed. There is no rush right now
But I'm happy to merge this PR and switch faro from fork to the upstream version of mimir-proxies

Please let me know if I can merge this PR

**Comment by leveyjam** [comment]
Hey - @grafana/app-o11y-receivers-squad aren't blocked by this, so we can take our time and come up with the right solution.
CC: @alexgreenbank @ywwg - https://raintank-corp.slack.com/archives/C046D0L50G4/p1689758772343879 - this is part of a broader proxy->proxy->proxy error propagation issue, where proxies interact in this string of incompatible http status code handling:

proxy chain error handling walk-through:
0) Load Balancer
1) Faro: inherits `mimir-proxies` middleware handling, can reject a request + return status before Faro code runs.
1a) Faro: return `202` or `4xx` based on our checks (we *do not* interpret 4xx/5xx any more).  
1z) Faro **used to do generic string->http status code to fix impossible to diagnose issues** https://github.com/grafana/app-o11y-kwl-endpoint/blob/main/pkg/exporter/utils/errormapper_test.go (this is no longer used - we are the 1st hop after the Load Balancer, and only inherit mimir-proxies checks + apply our own checks - we respond `202` and downstream errors are logged - not returned to the caller (we process async).
---- From Faro's perspective, at this point, we no longer care what happens, but anyone sending data to otlp-gateway likely does:
2) otlp-gateway: inherits `mimir-proxies` middleware handling, can reject a request + return status before otlp-gateway code runs.
2a) https://github.com/grafana/otlp-gateway/blob/602985d43b60ef0dc6729999563d57f1cc38c65c/pkg/logs/proxy.go#L177 (example of 4xx instead of 5xx) 
2b) otlp-gateway's own error mapping: https://github.com/grafana/otlp-gateway/blob/c529c8e99f06f2cd866af711e8bad7ffc8fef26d/pkg/traces/internal/error.go#L48 (this one is specific to Tempo downstream)

I haven't really scratched the surface here - but I don't think we should rush this. Faro are running @mar4uk's version with a fix that solves our problem (SLO Budget Burn accepting requests - Faro's clients are instrumented web applications which do not care about the response code - but that isn't a solution that works for all consumers of `mimir-proxies`, so we should settle on something sustainable for all).

File: pkg/server/middleware/request_limits.go
Comment (npazosmendez) [inline pkg/server/middleware/request_limits.go:43]: This is already being logged in the line above. Or are there any reasons to use the struct's logger instead of the span's from here?

File: pkg/server/middleware/request_limits.go
Comment (npazosmendez) [inline pkg/server/middleware/request_limits.go:60]: ditto

File: pkg/server/middleware/request_limits.go
Comment (mar4uk) [reply pkg/server/middleware/request_limits.go:43]: If I'm not mistaken spanlogger adds logs inside the span. 
So if the tracing is not working properly and spans are not sent we will not be able to see these logs (This is what happening in the faro collector right now, not all spans are sent).
That's why I added a logger here, to at least see these logs in Loki when I can't see them in Tempo

Comment (npazosmendez) [reply pkg/server/middleware/request_limits.go:43]: I'm not a tracing expert, but I think the logger from `spanlogger` logs like a regular logger. It [combines](https://github.com/grafana/dskit/blob/c30720760c5d23eef733a3e7e299e8247e438a24/spanlogger/spanlogger.go#L34) a span and a logger, and adds the traceID as a field. And the things [we log](https://github.com/grafana/transmog/blob/94f5dc4893a9ff08a2686116bead85019c09b2f5/pkg/graphite/querier/api/render.go#L65) we see them in [Loki](https://ops.grafana-ops.net/explore?panes=%7B%22NrH%22:%7B%22datasource%22:%22000000193%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bname%3D%5C%22graphite-querier%5C%22%7D%7C%3D%20%20%60render%20request%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22000000193%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%221689768047678%22,%22to%22:%221689768376397%22%7D%7D%7D&schemaVersion=1&orgId=1).

But I may be missing something here. Did you see this not working as expected?

**Review by leveyjam — COMMENTED** [review]
Sorry - never hit submit on the review 🤦

File: pkg/server/middleware/request_limits.go
Comment (leveyjam) [reply pkg/server/middleware/request_limits.go:43]: 👋 - I never hit "Submit" on my review, my take on this was that spanlogger is a utility that adds `elapsed` to the log line, rather than having anything to do with tracing (I saw "mimir/utils/spanlogger" and thought it must be a utility they've written to simplify calculating the "span" of time, but I could be wrong and it could be doing something with trace spans ).

Comment (npazosmendez) [reply pkg/server/middleware/request_limits.go:43]: I think that apart from that it's main purpose is to unify the creation and holding of both a span and a logger. That's why it's always followed by `defer log.Span.Finish()`

```go
log, ctx := spanlogger.New(r.Context(), "middleware.RequestLimits.Wrap")
defer log.Span.Finish()
```

Comment (mar4uk) [reply pkg/server/middleware/request_limits.go:43]: > And the things [we log](https://github.com/grafana/transmog/blob/94f5dc4893a9ff08a2686116bead85019c09b2f5/pkg/graphite/querier/api/render.go#L65) we see them in [Loki](https://ops.grafana-ops.net/explore?panes=%7B%22NrH%22:%7B%22datasource%22:%22000000193%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bname%3D%5C%22graphite-querier%5C%22%7D%7C%3D%20%20%60render%20request%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22000000193%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%221689768047678%22,%22to%22:%221689768376397%22%7D%7D%7D&schemaVersion=1&orgId=1).

Thank you for the example! Trying to find faro-collector logs the same way you do for graphite proxy, but I don't see logs from spanlogger, only logs from struct's logger. Need to debug and understand why it happens
[Example of faro-collector logs](https://ops.grafana-ops.net/explore?panes=%7B%22NrH%22:%7B%22datasource%22:%22000000193%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bname%3D%5C%22faro-collector%5C%22%7D%20%7C%3D%20%20%60failed%20to%20read%20request%20body%60%20%21%3D%20%5C%22middleware.RequestLimits.Wrap%5C%22%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22000000193%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%221689768047678%22,%22to%22:%221689768376397%22%7D%7D%7D&schemaVersion=1&orgId=1) (use query which filter out struct's log, they have prefix `middleware.RequestLimits.Wrap`)

Comment (leveyjam) [reply pkg/server/middleware/request_limits.go:43]: @mar4uk - I think I know what's going on.
```
level.Warn(log).Log("msg", "failed to read request body", "err", err)
```
^ This was added in a previous PR, and that `log` isn't logging anywhere. It wasn't connected to the Faro tracing issues, etc, I think it's just not going to be visible unless `l.logger` is used.

I think one-or-other of the `level.Warn(...).Log(...)` lines will work, but the using the spanlogger to give us the timings would be pretty useful when it's a timeout in middleware we're reporting.

Comment (leveyjam) [reply pkg/server/middleware/request_limits.go:43]: Graphite Proxies does this:
```
log.Logger = app.Logger
```
That log.Logger is from: "github.com/grafana/mimir/pkg/util/log"

```
var (
	// Logger is a shared go-kit logger.
	// TODO: Change all components to take a non-global logger via their constructors.
	// Prefer accepting a non-global logger as an argument.
	Logger = log.NewNopLogger()
)
```
So I'll see how our logging can be integrated into this mode of initialization, and we can forego passing a logger as an arg to middleware constructors 👍

Comment (leveyjam) [reply pkg/server/middleware/request_limits.go:43]: Thanks @npazosmendez - I discussed offline with @mar4uk and I've a little more free time right now, so I'll push a commit asap to remove the logger Faro was using, and we'll test Faro works via changes to how we initialize the context in our own time.
Thanks for the suggestion

File: pkg/server/middleware/request_limits.go
Comment (mar4uk) [reply pkg/server/middleware/request_limits.go:19]: I checked this repo and found that 499 is already used in pkg/errorx/http_translator.go
https://github.com/grafana/mimir-proxies/blob/6b2a4f93edfe67b44217edb1d6cf6cac196b3890/pkg/errorx/http_translator.go#L13
and in pkg/datadog/apiwrite/apiwrite.go
https://github.com/grafana/mimir-proxies/blob/6b2a4f93edfe67b44217edb1d6cf6cac196b3890/pkg/datadog/apiwrite/apiwrite.go#L26

So maybe we can keep 499 here as well?

Comment (leveyjam) [reply pkg/server/middleware/request_limits.go:18]: Nice find @mar4uk ! I think we have an issue where we point out that we need mimir-proxies to behave more like otlp-gateway, so keeping the 499 does exactly that - I think we're all agreed this is the way to standardise (499, instead of a 408 that's 50/50 a 4xx/5xx, etc).

Comment (leveyjam) [reply pkg/server/middleware/request_limits.go:18]: https://github.com/grafana/app-o11y-kwl-endpoint/issues/149#issue-1701072635

Comment (leveyjam) [reply pkg/server/middleware/request_limits.go:18]: (issue mentioning we need a 499 instead of 500 response, to mirror otlp-gateway)

File: pkg/server/middleware/request_limits.go
Comment (leveyjam) [reply pkg/server/middleware/request_limits.go]: @npazosmendez @jesusvazquez - I think this is resolved now. We've kept the spanlogger, and separately I will look at how to get Faro logging setup to use it (by copying the example from graphite-proxy, most likely).
But this PR is now ready for general consumption - thanks @mar4uk for doing the hardest part! (and thank you @jesusvazquez and @npazosmendez for the equally hard reviewing part!). 
🙇

**Comment by leveyjam** [comment]
Hey @mar4uk - I'm merging this (it's your PR so I wanted to let you know!)
Thanks everyone for the reviews 👍

**Pull request merged**
